### PR TITLE
Visual Instruction Presentation Customization

### PR DESCRIPTION
**Pull request by JThramer**
Implements #1503. 

* Adding delegate method to customize attributed string that is computed by `InstructionPresenter`.

To-do: 
- [ ] Add tests
- [ ] Add example

/cc @mapbox/navigation-ios

### CODE REVIEW

File: MapboxNavigation/NavigationViewController.swift [hunk 1]
```diff
@@ -538,6 +538,10 @@ extension NavigationViewController: RouteMapViewControllerDelegate {
         }
         return roadName
     }
+    
+    @objc public func label(_ label: InstructionLabel, willPresent instruction: VisualInstruction, as presented: NSAttributedString) -> NSAttributedString? {
```
Comment (bsudekum) [inline MapboxNavigation/NavigationViewController.swift]: Curious, why does NavigationViewController need to implement the delegate method?

Comment (JThramer) [reply MapboxNavigation/NavigationViewController.swift]: This is just plumbing the delegate method through the view-controller hierarchy, just like we do [here](https://github.com/mapbox/mapbox-navigation-ios/pull/1530/files#diff-ab5a908dc637e51fa78cf2be28d033ecL527) for a `NavigationMapViewDelegate` method or [here](https://github.com/mapbox/mapbox-navigation-ios/pull/1530/files#diff-292af689ce88d5b5aa06966456abc0e9R607) for the _same_ method (`label(_:willPresent:as:)`), being routed through `RouteMapViewController`.

It's not really accurate to say that the `NavigationViewController` is _implementing_ the delegate method so much as it's merely passing the message onto _its_ delegate.

**Comment by JThramer** [comment]
Demonstrative screenshot, where in the delegate method, I implemented logic to reverse the presented strings: 
simulator screen shot - iphone 8 plus - 2018-06-28 at 15 39 12

File: MapboxNavigation/InstructionLabel.swift [hunk 2]
```diff
@@ -28,10 +29,28 @@ open class InstructionLabel: StylableLabel, InstructionPresenterDataSource {
...
+    
+    /**
+     Called when an InstructionLabel will present a visual instruction.
+     
+     - parameter label: The label that the instruction will be presented on.
+     - parameter instruction: the `VisualInstruction` that will be presented.
+     - parameter presented: the formatted string that is provided by the instruction presenter
+     - returns: optionally, a customized NSAttributedString that will be presented instead of the default, or if nil, the default behavior will be used.
+     */
+    @objc optional func label(_ label: InstructionLabel, willPresent instruction: VisualInstruction, as presented: NSAttributedString) -> NSAttributedString?
```
Comment (1ec5) [inline MapboxNavigation/InstructionLabel.swift]: A delegate method’s first argument should always be the delegator. Is the delegator an InstructionLabel, or something else? Can it be either or two classes? That would be an antipattern in Cocoa.

Does this “will” method’s name make the implementation’s purpose clear enough? Is it clear that it’s an opportunity to override the presentation?

Comment (1ec5) [inline MapboxNavigation/InstructionLabel.swift]: The Objective-C signature of this method is currently `-label:willPresent:as:`. I think it should be `-…:willPresentVisualInstruction:label:presentedAttributedString:` or somesuch.

Comment (JThramer) [reply MapboxNavigation/InstructionLabel.swift]: Ah, right you are. Good catch. Will fix.

File: Examples/Swift/ViewController.swift [hunk 3]
```diff
@@ -442,6 +452,20 @@ extension ViewController: NavigationViewControllerDelegate {
     }
 }
 
+// Mark: VisualInstructionDelegate
+extension ViewController: VisualInstructionDelegate {
+    func label(_ label: InstructionLabel, willPresent instruction: VisualInstruction, as presented: NSAttributedString) -> NSAttributedString? {
+        guard exampleMode == .customInstruction else { return nil }
```
Comment (bsudekum) [inline Examples/Swift/ViewController.swift]: With `VoiceController(_:WillSpeak:Instruction:)`, returning nil omits the voice instruction. I'm guess that's not the case here? In reality, I can't really think of a use case where a developer would want to show nothing though.

**Comment by bsudekum** [comment]
With https://github.com/mapbox/mapbox-navigation-ios/pull/1514 coming down the line, can you think of anything that might make this play nicely with the lanes view? Maybe returning nil hides/omits the view? https://github.com/mapbox/mapbox-navigation-ios/pull/1530#discussion_r199961726

File: MapboxNavigationTests/InstructionsBannerViewIntegrationTests.swift [hunk 4]
```diff
@@ -69,6 +71,16 @@ class InstructionsBannerViewIntegrationTests: XCTestCase {
 
         super.tearDown()
     }
+    
+    func testCustomVisualInstructionDelegate() {
```
Comment (bsudekum) [inline MapboxNavigationTests/InstructionsBannerViewIntegrationTests.swift]: Also add a test for returning nil.

File: Examples/Swift/ViewController.swift (hunk 3 above)
Comment (JThramer) [reply Examples/Swift/ViewController.swift]: That's kind of the idea. The utility of "I only want to override default behavior under specific circumstances" seems to outweigh the utility of "it's nice to say 'show nothing' via `nil` instead of `NSAttributedString()`". In this _specific_ example you could _also_ implement default behavior by returning `presented`, but I think we should get into the general habit of using a  "returning nil means the delegate [has no answer](http://thecodelesscode.com/case/31) to the delegator, and thus default behavior should be used" pattern in our delegation, as otherwise you kind of force the developer to implement logic if they implement the optional delegate function. Not a super great pattern.

File: MapboxNavigationTests/InstructionsBannerViewIntegrationTests.swift (hunk 4 above)
Comment (JThramer) [reply MapboxNavigationTests/InstructionsBannerViewIntegrationTests.swift]: - [ ] Test for `nil` response

File: MapboxNavigationTests/InstructionsBannerViewIntegrationTests.swift [hunk 5]
```diff
@@ -69,6 +71,27 @@ class InstructionsBannerViewIntegrationTests: XCTestCase {
...
+        let view = instructionsView()
+        view.instructionDelegate = TextReversingDelegate()
```
Comment (frederoni) [inline MapboxNavigationTests/InstructionsBannerViewIntegrationTests.swift]: `instructionDelegate` will be deallocated immediately because it's weak. Keep a local ref to `TextReversingDelegate()`. (×2)
